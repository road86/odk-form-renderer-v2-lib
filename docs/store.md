# Walkthrough through Redux store

## Store

The store is initialized in `src/store.tsx` and added in `src/index.tsx` file.

## Reducer

Currently, there is one single reducer, that resides in the `src/ducks/formstate.ts`.

### State of the store
The state of the store can be best represented as below:

```js
export interface FormState {
  userInput: object;
  errors: string[];
  optionList: object;
  isFormSubmitted: boolean;
}
```
### Brief description of the major keys

#### userInput

The `userInput` key contains the most updated form field values.

#### errors

The `errors` hold the list of field names that have input values missing when required or violating form constraints. If the form successfully passes all the constraints or does not have any required missing values, the `errors` key will be an empty array with no field names. And form submission will only be successful if this key is empty.

#### optionList 

The `optionList` holds all the options of single select and multiple select presents in the forms at the time of rendered. It add or removes any options based on user input of fields.

#### isFormSubmitted

The initial value of this key is `false`. It acts as a utility flag on whether to show required or constraint messages If a form gets submitted and the submission is not successful, the flag is assigned `true`.