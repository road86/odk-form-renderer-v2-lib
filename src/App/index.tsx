import { library } from '@fortawesome/fontawesome-svg-core';
import { faMinusCircle, faPlusCircle } from '@fortawesome/free-solid-svg-icons';
import * as React from 'react';
import { connect } from 'react-redux';
import { Button, Col, Container, Row } from 'reactstrap';
import { Store } from 'redux';
import KbAlert from '../components/Alert';
import DropDown from '../components/DropDown';
import GroupTypeEvaluator from '../components/typeEvalutors/Group';
import {
  getUserInputFromStore,
  isErrorsArrayEmpty,
  setUserInputObj,
} from '../store/ducks/formState';

library.add(faPlusCircle, faMinusCircle);

export interface AppProps {
  csvList: any;
  isNoErrors: any;
  userInputObj: any;
  userInputJson: any;
  defaultLanguage: string;
  formTitle: string;
  fieldElements: any;
  setUserInputAction: typeof setUserInputObj;
  languageOptions: any;
  handleSubmit(userInput: any): any;
}

export interface AppState {
  defaultLanguage: string;
  isSubmissionError: boolean;
}

class App extends React.Component<AppProps, AppState> {
  constructor(props: AppProps) {
    super(props);
  }

  public componentDidMount() {
    const { userInputJson, userInputObj } = this.props;
    if (userInputJson && userInputJson !== userInputObj) {
      this.props.setUserInputAction(userInputJson);
    }
    this.setState({
      defaultLanguage: this.props.defaultLanguage,
      isSubmissionError: false,
    });
  }

  public handleSelect = (languageName: string) => {
    this.setState({ defaultLanguage: languageName });
  };

  public render() {
    const { csvList, fieldElements, formTitle, languageOptions } = this.props;
    const { defaultLanguage } = this.state || this.props;

    const props = {
      csvList,
      defaultLanguage,
      fieldElements,
      fieldParentTreeName: '',
      languageOptions,
    };

    return (
      <Container className="form-container">
        <Row className={'form-title formTitle'}>
          <Col>
            <h3 className="headerText">{formTitle}</h3>
          </Col>
          <DropDown
            languages={...languageOptions}
            onChangeSelect={this.handleSelect}
          />
        </Row>
        {this.state && this.state.isSubmissionError && (
          <KbAlert
            color={'danger'}
            isOpen={this.state.isSubmissionError}
            handleToggle={this.toggleStateValue}
            headerText={'Oh snap! You got an error!'}
            bodyText={
              'Please make sure the required fields are not missing and there are no errors'
            }
          />
        )}
        <Row className="formFieldBody">
          <Col>
            <GroupTypeEvaluator {...props} />
            <Row className="welcome-box">
              <Col>
                <Button className="btn btn-success" onClick={this.handleClick}>
                  Submit
                </Button>
              </Col>
            </Row>
          </Col>
        </Row>
      </Container>
    );
  }

  // tslint:disable-next-line: variable-name
  private handleClick = (_event: React.MouseEvent<HTMLButtonElement>) => {
    const { handleSubmit, isNoErrors, userInputObj } = this.props;
    handleSubmit('submitted');
    if (isNoErrors) {
      handleSubmit(userInputObj);
    } else {
      handleSubmit('Field Violated');
      this.setState({ isSubmissionError: true });
      window.scrollTo(0, 0);
    }
  };

  private toggleStateValue = () => {
    this.setState({ isSubmissionError: false });
  };
}

/** connect the component to the store */

/** Interface to describe props from mapStateToProps */
interface DispatchedStateProps {
  isNoErrors: any;
  userInputObj: any;
}

/** Map props to state  */
const mapStateToProps = (state: Partial<Store>): DispatchedStateProps => {
  const result = {
    isNoErrors: isErrorsArrayEmpty(state),
    userInputObj: getUserInputFromStore(state),
  };
  return result;
};

/** map props to actions */
const mapDispatchToProps = {
  setUserInputAction: setUserInputObj,
};

/** connect Decimal component to the redux store */
const ConnectedApp = connect(
  mapStateToProps,
  mapDispatchToProps
)(App);

export default ConnectedApp;
