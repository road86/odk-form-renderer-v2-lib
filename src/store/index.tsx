import { createStore, Store } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import reducer, { FormState } from './ducks/formState';

/** The initial store */
const store: Store & FormState = createStore(reducer, composeWithDevTools());

export default store;
