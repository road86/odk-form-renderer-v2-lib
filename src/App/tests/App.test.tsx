import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import App from '../../App';
import store from '../../store';

describe('App', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders App without crashing', () => {
    const mock: any = jest.fn();
    const div = document.createElement('div');
    const props = {
      choices: {},
      csvList: {},
      defaultLanguage: 'English',
      fieldElements: [],
      formTitle: 'ODK',
      handleSubmit: mock,
      languageOptions: [
        {
          label: 'English',
          value: 'English',
        },

        {
          label: 'Bangla',
          value: 'Bangla',
        },
      ],
      userInputJson: {},
    };
    ReactDOM.render(
      <Provider store={store}>
        <App {...props} />
      </Provider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders App correctly', () => {
    const mock: any = jest.fn();
    const props = {
      choices: {},
      csvList: {},
      defaultLanguage: 'English',
      fieldElements: [],
      formTitle: 'ODK',
      handleSubmit: mock,
      languageOptions: [
        {
          label: 'English',
          value: 'English',
        },
        {
          label: 'Bangla',
          value: 'Bangla',
        },
      ],
      userInputJson: {},
    };
    const wrapper = mount(
      <Provider store={store}>
        <App {...props} />
      </Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
