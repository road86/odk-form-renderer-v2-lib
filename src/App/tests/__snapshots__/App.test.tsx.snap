// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders App correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(App)
    csvList={Object {}}
    defaultLanguage="English"
    fieldElements={Array []}
    formTitle="ODK"
    handleSubmit={[MockFunction]}
    userInputJson={Object {}}
  >
    <App
      csvList={Object {}}
      defaultLanguage="English"
      fieldElements={Array []}
      formTitle="ODK"
      handleSubmit={[MockFunction]}
      isNoErrors={true}
      setUserInputAction={[Function]}
      userInputJson={Object {}}
      userInputObj={Object {}}
    >
      <Container
        className="form-container"
        tag="div"
      >
        <div
          className="form-container container"
        >
          <Row
            className="form-title formTitle"
            tag="div"
          >
            <div
              className="form-title formTitle row"
            >
              <Col
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <div
                  className="col"
                >
                  <h3>
                    ODK
                  </h3>
                </div>
              </Col>
            </div>
          </Row>
          <Row
            className="formFieldBody"
            tag="div"
          >
            <div
              className="formFieldBody row"
            >
              <GroupTypeEvaluator
                csvList={Object {}}
                defaultLanguage="English"
                fieldElements={Array []}
                fieldParentTreeName=""
              >
                <div />
              </GroupTypeEvaluator>
            </div>
          </Row>
          <Row
            className="welcome-box"
            tag="div"
          >
            <div
              className="welcome-box row"
            >
              <Col
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <div
                  className="col"
                >
                  <Button
                    className="btn btn-success"
                    color="secondary"
                    onClick={[Function]}
                    tag="button"
                  >
                    <button
                      aria-label={null}
                      className="btn btn-success btn btn-secondary"
                      onClick={[Function]}
                      type="button"
                    >
                      Submit
                    </button>
                  </Button>
                </div>
              </Col>
            </div>
          </Row>
        </div>
      </Container>
    </App>
  </Connect(App)>
</Provider>
`;
